src/view.c

```diff
@@ -857,51 +907,252 @@ view_on_output_destroy(struct view *view)
+		} else if (!strcasecmp(direction, "down")) {
+			// top edge to top/bottom edges
+			_view_get_next_edges(view, direction,
+				em.top_edge, em.down_max_dy, 0, rc.gap, &near, &far);
+			dy = MIN(near, far);
+			// bottom edge to top edge
+			_view_get_next_edges(view, direction,
+				em.bottom_edge, em.down_max_dy, -rc.gap, 0, &near, &far);
+			dy = MIN(dy, near);
+		} else {
+			wlr_log(WLR_ERROR, "invalid direction: %s", direction);
+			return;
+		}
+	} else if (!strcasecmp(snap, "screen")) {
+		if (!strcasecmp(direction, "left")) {
+			dx = em.left_max_dx;
+		} else if (!strcasecmp(direction, "up")) {
+			dy = em.up_max_dy;
+		} else if (!strcasecmp(direction, "right")) {
+			dx = em.right_max_dx;
+		} else if (!strcasecmp(direction, "down")) {
+			dy = em.down_max_dy;
+		} else {
+			wlr_log(WLR_ERROR, "invalid direction: %s", direction);
+			return;
+		}
+	} else {
+		wlr_log(WLR_ERROR, "invalid snap: %s", snap);
+		return;
 	}
-	if (usable.width == output->wlr_output->width
-			&& output->wlr_output->scale != 1) {
-		usable.width /= output->wlr_output->scale;
+
+	view_move(view,
+		dx ? view->current.x + dx : view->pending.x,
+		dy ? view->current.y + dy : view->pending.y);
+}
+
+void
+view_grow_to_edge(struct view *view, const char *direction)
+{
+	assert(view);
+	if (view->fullscreen || view->maximized)
+		return;
+
+	struct output *output = view->output;
+	if (!output_is_usable(output)) {
+		wlr_log(WLR_ERROR, "view has no output, not growing view");
+		return;
+	}
+	if (!direction) {
+		wlr_log(WLR_ERROR, "invalid direction, not growing view");
+		return;
+	}
+
+	struct edge_manip_hints em = view_get_edge_manip_hints(view);
+	struct wlr_box geo = view->pending;
+	int dx = 0, dy = 0;
+	int near, far;
+	if (!strcasecmp(direction, "left")) {
+		// left edge to left/right edges
+		_view_get_next_edges(view, direction,
+			em.left_edge, em.left_max_dx, rc.gap, 0, &near, &far);
+		dx = MAX(near, far);
+		geo.width += -dx;
+		geo.x     += dx;
+	} else if (!strcasecmp(direction, "up")) {
+		// top edge to top/bottom edges
+		_view_get_next_edges(view, direction,
+			em.top_edge, em.up_max_dy, rc.gap, 0, &near, &far);
+		dy = MAX(near, far);
+		geo.height += -dy;
+		geo.y      += dy;
+	} else if (!strcasecmp(direction, "right")) {
+		// right edge to left/right edges
+		_view_get_next_edges(view, direction,
+			em.right_edge, em.right_max_dx, -rc.gap, 0, &near, &far);
+		dx = MIN(near, far);
+		geo.width  += dx;
+	} else if (!strcasecmp(direction, "down")) {
+		// bottom edge to top/bottom edges
+		_view_get_next_edges(view, direction,
+			em.bottom_edge, em.down_max_dy, -rc.gap, 0, &near, &far);
+		dy = MIN(near, far);
+		geo.height += dy;
+	} else {
+		wlr_log(WLR_ERROR, "invalid direction: %s", direction);
+		return;
 	}
 
-	int x = 0, y = 0;
+	view_move_resize(view, geo);
+}
+
+void
+view_shrink_to_edge(struct view *view, const char *direction)
+{
+	assert(view);
+	if (view->fullscreen || view->maximized)
+		return;
```

Comment:
Please always use curly braces for if statements (and for loops, ..): It is way too easy to add another line to the body during debugging or similar which then can blow up everything.

@johanmalm 
We should really try to get this into checkpatch.pl as well so the CI would warn about it.

Reply:
Fixed in: 75c2d4c43bac1eaa07c96e037118d170e64f2a52 (hoping I caught all of them)